### PR TITLE
Add support for UPDATE_COM_METHOD

### PR DESCRIPTION
**Pull request by Patrik-Stas**

### CODE REVIEW

File: vcxagency-client/src/agency-flows.js [hunk 1]
```diff
@@ -137,6 +138,23 @@ async function vcxFlowUpdateMsgsFromAgent (clientWh, sendToAgency, recipientAgen
...
+ * @param {string} clientWh - wallet handle of agency client
+ * @param {object} sendToAgency - function which passes final message to an agency
+ * @param {string} clientVkey - verkey the client uses to talk to Agency
+ * @param {string} recipientAgentDid - DID of the Agent
+ * @param {string} recipientAgentVerkey - Verkey of the Agent
+ * @param {string} webhookUrl - url where Agent's notifications shall be sent
+ */
+async function vcxFlowSetWebhookUrl (clientWh, sendToAgency, recipientAgentDid, recipientAgentVerkey, clientVkey, webhookUrl) {
+  const encryptionOurs = clientVkey
+  // todo: seems like the failus in agent decryption were because i am using here pack, instead of packAsUtf8, see above functions
```
Comment (mirgee) [inline vcxagency-client/src/agency-flows.js:152]: It seems like this comment is not needed anymore :)

File: vcxagency-client/src/agency-flows.js [hunk 2]
```diff
@@ -137,6 +138,23 @@ async function vcxFlowUpdateMsgsFromAgent (clientWh, sendToAgency, recipientAgen
...
+ Update statusCode of messages specified by UIDs per Agent-Connections
+ * @param {string} clientWh - wallet handle of agency client
+ * @param {object} sendToAgency - function which passes final message to an agency
+ * @param {string} clientVkey - verkey the client uses to talk to Agency
+ * @param {string} recipientAgentDid - DID of the Agent
+ * @param {string} recipientAgentVerkey - Verkey of the Agent
+ * @param {string} webhookUrl - url where Agent's notifications shall be sent
+ */
+async function vcxFlowSetWebhookUrl (clientWh, sendToAgency, recipientAgentDid, recipientAgentVerkey, clientVkey, webhookUrl) {
+  const encryptionOurs = clientVkey
```
Comment (mirgee) [inline vcxagency-client/src/agency-flows.js:151]: Personally, I wouldn't waste lines with renaming the parameters (in this file)

File: vcxagency-client/src/agency-flows.js [hunk 3]
```diff
@@ -16,6 +16,7 @@
 
 'use strict'
 
+const { buildMsgVcxV2UpdateWebhookUrl } = require('./messaging/client2agency/msgs-agent')
```
Comment (mirgee) [inline vcxagency-client/src/agency-flows.js:19]: Why V2? Do we need to reference dummy-cloud agency here?

File: vcxagency-client/src/messaging/client2agency/msgs-agent.js [hunk 4]
```diff
@@ -82,6 +84,24 @@ function buildMsgVcxV2MsgStatusUpdatedByConns (failedUidsByConns, updatedUidsByC
   return msg
 }
 
+function buildMsgVcxV2UpdateWebhookUrl (webhookUrl) {
+  const msg = {
+    '@type': MSGTYPE_UPDATE_COM_METHOD,
+    id: '123',
```
Comment (mirgee) [inline vcxagency-client/src/messaging/client2agency/msgs-agent.js:90]: Maybe refer to dummy-cloud-agency impl. here

File: vcxagency-node/test/unit/entities/agent/agent.spec.js [hunk 5]
```diff
@@ -82,4 +82,44 @@ describe('agent operations', () => {
...
+
+    expect(await agentAo.getWebhook()).toBe(null)
+  })
+
+  it('should throw if webhook is not http or https', async () => {
+    const { agentDid: agentDidOnCreate } = await createAgentData(clientDid, clientVerkey, serviceIndyWallets, serviceStorage)
+    const entityRecord = await serviceStorage.loadEntityRecord(agentDidOnCreate)
+    const agentAo = await buildAgentAO(entityRecord, serviceIndyWallets, serviceStorage)
+
+    // act
```
Comment (mirgee) [inline vcxagency-node/test/unit/entities/agent/agent.spec.js:114]: Unnecessary comment (below another one)

File: vcxagency-node/test/unit/storage/pgstorage-webhooks.spec.js [hunk 6]
```diff
@@ -45,4 +45,32 @@ describe('storage', () => {
     expect(await storage.getAgentWebhook(A)).toBe('https://example.org')
     expect(await storage.getAgentWebhook(B)).toBe('http://localhost:9123')
   })
+
+  it('should return null if webhook was unset', async () => {
+    const A = uuid.v4()
```
Comment (mirgee) [inline vcxagency-node/test/unit/storage/pgstorage-webhooks.spec.js:50]: I like how thoroughly tested this func is, but way more tests can fit in here without all these comments and emty lines :)

File: vcxagency-node/src/service/entities/agent/agent.js [hunk 7]
```diff
@@ -151,23 +155,33 @@ async function buildAgentAO (entityRecord, serviceWallets, serviceStorage, route
...
+      throw Error(`${whoami} VCX V2 Forward message expected to have field '@msg'. Message: ${JSON.stringify(msgObject)}'`)
     }
     return router.routeMsg(fwd, Buffer.from(JSON.stringify(msg)))
   }
 
+  async function _handleUpdateComMethod (msgObject) {
+    const { comMethod: { value, type, id } } = msgObject
+    if (type === '2') {
+      await setWebhook(value)
+      return buildMsgCommMethodUpdated(id) // TODO: What's meaning of 'id' field? This reflects dummy-cloud-agency impl.
```
Comment (mirgee) [inline vcxagency-node/src/service/entities/agent/agent.js:170]: It's good to refer there, but I think the function definition is a better place for it than this one.

File: vcxagency-node/test/unit/storage/pgstorage-webhooks.spec.js (hunk 6 above)
Comment (Patrik-Stas) [reply vcxagency-node/test/unit/storage/pgstorage-webhooks.spec.js:50]: I've deleted commented but I'll keep empty lines. When possible, I like to split tests in 3 parts - setup, execute whats being test, verify results

File: vcxagency-node/src/service/entities/agent/agent.js (hunk 7 above)
Comment (Patrik-Stas) [reply vcxagency-node/src/service/entities/agent/agent.js:170]: I found it important to note that this particular method reflects dummy cloud agency behaviour - it takes received `id` and put it into `id` in response as well.

**Pull request merged**